zkvm/src/constraints.rs

```diff
@@ -44,6 +45,42 @@ pub struct CommitmentWitness {
     blinding: Scalar,
 }
 
+impl Constraint {
+    pub fn verify<CS: r1cs::ConstraintSystem>(self, cs: &mut CS) -> Result<(), VMError> {
+        cs.specify_randomized_constraints(move |cs| {
+            let z = cs.challenge_scalar(b"verify challenge");
+
+            // Flatten the constraint into one expression
+            // Note: cloning because we can't move out of captured variable in an `Fn` closure
+            let expr = self.clone().flatten(cs, z);
+
+            // Add the resulting expression to the constraint system
+            cs.constrain(expr.to_r1cs_lc());
+
+            Ok(())
+        })
+        .map_err(|e| VMError::R1CSError(e))
+    }
+
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS, z: Scalar) -> Expression {
+        match self {
+            Constraint::Eq(expr1, expr2) => expr1 - expr2,
+            Constraint::And(c1, c2) => {
+                let a = c1.flatten(cs, z);
```

Comment:
i'm not sure you can use the same challenge across all levels of a boolean function. IMHO it's safer to generate it per each `and(a,b)` pair.

Reply:
ah that's true, fixed!